### ISSUE TITLE
Training Loss for Language Models

### ISSUE THREAD

**Issue by I-Good-Vegetable** [pomonam/jax-influence#2]

**Comment by I-Good-Vegetable** [comment]
Hi Dr. Bae,

Thanks for contributing the great paper. I have recently been reading your paper, ‘Studying Large Language Model Generalization with Influence Functions’. I noticed that the training loss of language models is different from other discriminative models. I'm wondering whether my implementation correct? I'm also wondering whether I should use same training loss for estimating iHVP and calculating training sample gradient.

Many thanks indeed!

Here is my implementation. I'm using GPT2

    def train_outputs(self, model: Module, batch: Any):
        # batch_size needs to be 1
        batch.pop('labels')
        logits = model(**batch).logits
        probs = logits.softmax(-1)
        pseudo_labels = torch.multinomial(probs[0], 1).view(1, -1)
        pseudo_labels = torch.roll(pseudo_labels, 1, -1)
        batch['labels'] = pseudo_labels
        return logits

    def train_loss_on_outputs(self, outputs, batch):
        # batch_size needs to be 1
        labels = batch['labels']
        shift_logits = outputs[..., :-1, :].contiguous()
        shift_labels = labels[..., 1:].contiguous()
        loss = CrossEntropyLoss(reduction='sum')(
            shift_logits.view(-1, shift_logits.size(-1)),
            shift_labels.view(-1),
        )
        return loss

**Comment by I-Good-Vegetable** [comment]
or should I take the argmax of model's prediction

**Comment by pomonam** [comment]
Thank you for the question! You should use sampled gradients for computing the EK-FAC factors (e.g., Lambda factor or pseudo-gradient covariance) but use the not-sampled gradient for computing influence scores. It is difficult to see if the implementation is correct with the given code, but if you can share end-to-end example, I can give more detailed comments!

**Comment by I-Good-Vegetable** [comment]
Hi Dr. Bae,

Thanks for the response. I was using torch-influence repo. Below is my code for the objective class and I'm using LiSSA to estimating ihvp.

```python
class LMObject(BaseObjective):
    def __init__(self, tokenizer, test_length) -> None:
        self.tokenizer = tokenizer
        self.test_length = test_length

    def train_outputs(self, model: Module, batch: Any):
        # batch_size needs to be 1
        logits = model(**batch).logits
        return logits

    def train_loss_on_outputs(
            self, 
            outputs, 
            batch,
            pseudo_loss: bool,
    ):
        labels = batch['input_ids'].clone()
        shift_logits = outputs[..., 1:-1, :].contiguous()
        shift_labels = labels[..., 2:].contiguous()
        mask = (shift_labels != self.tokenizer.pad_token_id).view(-1)
        losses = CrossEntropyLoss(reduction='none')(
            shift_logits.view(-1, shift_logits.size(-1)),
            shift_labels.view(-1),
        )
        loss = losses[mask].sum()
        return loss
    
    train_regularization = None

    def train_loss(
            self, 
            model: Module, 
            params: torch.Tensor, 
            batch: Any,
            pseudo_loss: bool,
    ) -> torch.Tensor:
        # batch_size needs to be 1
        outputs = self.train_outputs(model, batch)
        loss = self.train_loss_on_outputs(outputs, batch, pseudo_loss)
        return loss

    def test_loss(
            self, 
            model: Module, 
            params: torch.Tensor, 
            batch: Any,
    ) -> torch.Tensor:
        # batch_size needs to be 1
        labels = batch.pop('labels')
        outputs = model(**batch)
        logits = outputs.logits
        shift_logits = logits[..., :-1, :].contiguous()
        shift_labels = labels[..., 1:].contiguous()
        mask = (shift_labels != self.tokenizer.pad_token_id).view(-1)
        print(self.tokenizer.decode(shift_labels[..., mask][0]), self.tokenizer.decode(shift_logits.argmax(-1)[..., mask][0]))
        losses = CrossEntropyLoss(reduction='none')(
            shift_logits.view(-1, shift_logits.size(-1)),
            shift_labels.view(-1),
        )
        loss = losses[mask].sum()
        batch['labels'] = labels
        return loss
```

Many thanks indeed!

**Comment by pomonam** [comment]
No problem! I recently released the repo for computing influence functions with the WikiText example: https://github.com/pomonam/kronfluence. You can check it out to compare with this code.
